File: backend-fastapi/lib/clients/storage.py
import os
import uuid
from datetime import UTC, datetime, timedelta
from typing import cast
from urllib.parse import quote

from google.api_core.exceptions import Conflict, NotFound
from google.auth.credentials import AnonymousCredentials
from google.cloud.storage import Client
from google.oauth2 import service_account
from starlette.datastructures import UploadFile

from ..types_ import FileToUpload


class CloudStorageConfig:
    def __init__(
        self,
        project_id: str,
        bucket_name: str,
        access_expiration: int = 3600,
        credentials_file: str = "",
        emulator_public_url: str = "",
        emulator_private_url: str = "",
    ) -> None:
        self.project_id = project_id
        self.bucket_name = bucket_name
        self.access_expiration = access_expiration
        self.credentials_file = credentials_file
        self.emulator_public_url = emulator_public_url
        self.emulator_private_url = emulator_private_url


class CloudStorage:
    def __init__(self, config: CloudStorageConfig) -> None:
        # Parameters
        self.project_id: str = config.project_id
        self.bucket_name: str = config.bucket_name
        self.blob_access_expiration: int = config.access_expiration
        self.credentials_file: str = config.credentials_file
        self.emulator_public_url: str = config.emulator_public_url
        self.emulator_private_url: str = config.emulator_private_url

        # Storage client
        self.storage = Client(
            project=self.project_id,
            credentials=self.credentials,
            client_options=self.client_options,
        )

        # Initilaize the emulator
        if self.is_emulator:
            self._init_emulator()

        # Select the bucket
        self.bucket = self.storage.bucket(self.bucket_name)

    @property
    def is_emulator(self) -> bool:
        return bool(self.emulator_private_url and self.emulator_public_url)

    @property
    def credentials(
        self,
    ) -> service_account.Credentials | AnonymousCredentials | None:
        if self.is_emulator:
            return AnonymousCredentials()

        if self.credentials_file and os.path.exists(self.credentials_file):
            return service_account.Credentials.from_service_account_file(
                self.credentials_file
            )
        return None

    @property
    def client_options(self) -> dict | None:
        if self.is_emulator:
            return dict(api_endpoint=self.emulator_private_url)
        return None

    def _init_emulator(self) -> None:
        try:
            self.storage.create_bucket(self.bucket_name)
            print(f"Created bucket {self.bucket_name} in emulator")
        except Conflict:
            # error status code 409 = bucket exists
            pass

    def _get_emulator_file_url(self, filename: str) -> str:
        api_url = f"{self.emulator_public_url}/download/storage/v1"
        bucket_url = f"{api_url}/b/{self.bucket_name}"
        download_url = f"{bucket_url}/o/{quote(filename)}?alt=media"
        return download_url

    def get_signed_url(
        self, filename: str, expiration: int | None = None
    ) -> str:
        if self.is_emulator:
            return self._get_emulator_file_url(filename)

        expiration = expiration or self.blob_access_expiration
        blob = self.bucket.blob(filename)

        url = blob.generate_signed_url(
            version="v4",
            expiration=datetime.now(UTC) + timedelta(seconds=expiration),
            method="GET",
        )
        return url

    def upload_file(
        self,
        param: FileToUpload | UploadFile | str | None,
        destination: str = "",
    ) -> str:
        if not param:
            return ""

        if isinstance(param, str):
            file = FileToUpload.from_path(param)
        elif isinstance(param, UploadFile):
            file = FileToUpload.from_upload_file(param)
        else:
            file = param

        ext = os.path.splitext(file.name)[1]
        base_filename = os.path.splitext(destination or file.name)[0]
        filename = f"{base_filename}_{uuid.uuid4().hex}{ext}".lower()

        blob = self.bucket.blob(filename)
        blob.upload_from_string(file.buffer, content_type=file.mimetype)
        return cast(str, blob.name)

    def delete_file(self, filename: str) -> bool:
        try:
            blob = self.bucket.blob(filename)
            blob.delete()
            return True
        except NotFound:
            # 404 error when the filename does not exist
            return False

File: backend-express/src/lib/clients/storage.ts
import {
    Bucket,
    SaveOptions,
    Storage,
    StorageOptions,
} from "@google-cloud/storage";
import { randomUUID } from "crypto";
import { existsSync } from "fs";
import path from "path";

import { env } from "@/config";
import { FileToUpload } from "@/types";

interface GCSConfig {
    GCP_PROJECT_ID: string;
    GCS_BUCKET_NAME: string;
    GCS_BLOB_ACCESS_EXPIRATION?: number;
    GOOGLE_APPLICATION_CREDENTIALS?: string;
    GCS_EMULATOR_PRIVATE_URL?: string;
    GCS_EMULATOR_PUBLIC_URL?: string;
}

export class CloudStorage {
    private readonly projectId: string;
    private readonly bucketName: string;
    private readonly urlExpiration: number;
    private readonly emulatorPublicUrl?: string;
    private readonly emulatorPrivateUrl?: string;
    private readonly credentialsFile?: string;

    private readonly storage: Storage;
    private readonly bucket: Bucket;

    constructor(config: GCSConfig) {
        // Setting base configuration
        this.projectId = config.GCP_PROJECT_ID;
        this.bucketName = config.GCS_BUCKET_NAME;
        this.urlExpiration = config.GCS_BLOB_ACCESS_EXPIRATION || 3600;
        this.emulatorPublicUrl = config.GCS_EMULATOR_PUBLIC_URL;
        this.emulatorPrivateUrl = config.GCS_EMULATOR_PRIVATE_URL;
        this.credentialsFile = config.GOOGLE_APPLICATION_CREDENTIALS;

        // Creating the storage object
        const options = this.getStorageOptions();
        this.storage = new Storage(options);

        // Initialize the bucket in the emulator if needed
        if (this.isEmulator) {
            this.initEmulator().catch((err) => {
                console.error("Failed to initialize emulator:", err);
            });
        }

        // Select the bucket
        this.bucket = this.storage.bucket(this.bucketName);
    }

    get isEmulator(): boolean {
        return !!this.emulatorPrivateUrl && !!this.emulatorPublicUrl;
    }

    private getStorageOptions(): StorageOptions {
        const baseConfig: StorageOptions = { projectId: this.projectId };

        if (this.isEmulator) {
            return {
                ...baseConfig,
                apiEndpoint: `${this.emulatorPrivateUrl}`,
            };
        }

        if (this.credentialsFile && existsSync(this.credentialsFile)) {
            return {
                ...baseConfig,
                keyFilename: this.credentialsFile,
            };
        }

        return baseConfig;
    }

    private async initEmulator(): Promise<void> {
        try {
            await this.storage.createBucket(this.bucketName);
            console.log(`Created bucket ${this.bucketName} in emulator`);
        } catch (err) {
            // error status code 409 = bucket exists
            if ((err as any).code !== 409) throw err;
        }
    }

    private getEmulatorFileUrl(filename: string): string {
        return `${this.emulatorPublicUrl}/download/storage/v1/b/${
            this.bucketName
        }/o/${encodeURIComponent(filename)}?alt=media`;
    }

    public async getSignedUrl(
        filename: string,
        expiration?: number
    ): Promise<string> {
        if (this.isEmulator) {
            return this.getEmulatorFileUrl(filename);
        }

        expiration = expiration || this.urlExpiration;
        const file = this.bucket.file(filename);
        const [url] = await file.getSignedUrl({
            version: "v4",
            action: "read",
            expires: Date.now() + expiration * 1000,
        });
        return url;
    }

    public async uploadFile(
        file: FileToUpload | string | null,
        destination?: string
    ): Promise<string> {
        if (!file) {
            return "";
        }

        if (typeof file === "string") {
            file = FileToUpload.fromPath(file);
        }

        const ext = path.extname(file.originalname);
        const baseFilename = path.basename(
            destination || file.originalname,
            ext
        );
        const filename = `${baseFilename}_${randomUUID()}${ext}`.toLowerCase();
        const blob = this.bucket.file(filename);
        const metadata: SaveOptions = {
            contentType: file.mimetype,
            metadata: { contentType: file.mimetype },
        };
        await blob.save(file.buffer, metadata);
        return blob.name;
    }

    public async deleteFile(filename: string): Promise<boolean> {
        try {
            const blob = this.bucket.file(filename);
            await blob.delete();
            return true;
        } catch (error) {
            // 404 error when the filename does not exists
            if ((error as any).code !== 404) {
                throw error;
            }
            return false;
        }
    }
}

export const storage = new CloudStorage(env);
